src/OsPath/Internal/Include.hs

```diff
@@ -0,0 +1,335 @@
+--     PLATFORM_PATH = PosixPath | WindowsPath
+--     PLATFORM_PATH_SINGLE = 'PosixPath' | 'WindowsPath'
+--     IS_WINDOWS = 0 | 1
+
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE QuasiQuotes         #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving  #-}
+{-# LANGUAGE TemplateHaskell     #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- | Internal types and functions.
+
+module OsPath.Internal.PLATFORM_NAME
+  ( -- * The Path type
+    Path(..)
+  , toOsPath
+
+    -- * Validation functions
+  , isValidAbsDir
+  , isValidAbsFile
+  , isValidRelDir
+  , isValidRelFile
+  , hasParentDir
+
+    -- * Normalizing functions
+  , normalizeLeadingSeps
+  , normalizeTrailingSeps
+  , normalizeAllSeps
+#if IS_WINDOWS
+  , normalizeWindowsSeps
+#endif
+  , normalizeDrive
+  , normalizeDir
+  , normalizeFile
+
+    -- * Other helper functions
+  , extSep
+  , pathSep
+  , relRoot
+  , isWindows
+  )
+  where
+
+import Control.DeepSeq (NFData (..))
+import Control.Exception (displayException)
+import Data.Aeson (ToJSON (..), ToJSONKey(..))
+import Data.Aeson.Types (toJSONKeyText)
+import qualified Data.Text as Text (pack)
+import GHC.Generics (Generic)
+import Data.Data
+import Data.Hashable
+import qualified Language.Haskell.TH.Syntax as TH
+import System.OsPath.PLATFORM_NAME (PLATFORM_PATH)
+import qualified System.OsPath.PLATFORM_NAME as OsPath
+import System.OsString.Internal.Types (PLATFORM_STRING(..))
+import qualified System.OsString.PLATFORM_NAME as OsString
+
+-- | Path of some base and type.
+--
+-- The type variables are:
+--
+--   * @b@ — base, the base location of the path; absolute or relative.
+--   * @t@ — type, whether file or directory.
+--
+-- Internally it is a PLATFORM_PATH_SINGLE, which can be of two formats only:
+--
+-- 1. File format: @file.txt@, @foo\/bar.txt@, @\/foo\/bar.txt@
+-- 2. Directory format: @foo\/@, @\/foo\/bar\/@
+--
+-- All directories end in a trailing separator. There are no duplicate
+-- path separators @\/\/@, no @..@, no @.\/@, no @~\/@, etc.
+newtype Path b t = Path PLATFORM_PATH
+  deriving (Data, Typeable, Generic)
+
+-- | String equality.
+--
+-- The following property holds:
+--
+-- @show x == show y ≡ x == y@
+instance Eq (Path b t) where
+  (==) (Path x) (Path y) = x == y
+  {-# INLINE (==) #-}
+
+-- | String ordering.
+--
+-- The following property holds:
+--
+-- @show x \`compare\` show y ≡ x \`compare\` y@
+instance Ord (Path b t) where
+  compare (Path x) (Path y) = compare x y
+  {-# INLINE compare #-}
+
+-- | Same as 'show . OsPath.toOsPath'.
+--
+-- The following property holds:
+--
+-- @x == y ≡ show x == show y@
+instance Show (Path b t) where
+  show = show . toOsPath
+  {-# INLINE show #-}
+
+instance NFData (Path b t) where
+  rnf (Path x) = rnf x
+  {-# INLINE rnf #-}
+
+instance ToJSON (Path b t) where
+  toJSON =
+      either (error . displayException) toJSON
+    . OsPath.decodeUtf
```

Comment:
We can avoid the `error`, but it's up to debate:

* try `decodeUtf` (equals utf8 on unix and utf16le on windows)
* in case of failure, fall back to `latin1` on unix and `ucs-2` on windows, which both cannot fail (morally... the types of `decodeWith` lie in this case)

This is what I dislike about aeson. You can't really specify different encodings for the same type, unlike waargonaut.

Reply:
Sounds fine to me.
@hasufell Does your comment also apply to the encoding functions? Also, I found `System.OsString.Encoding.ucs2le`, but no `TextEncoding` for `latin1`.

Reply:
https://hackage.haskell.org/package/base-4.20.0.1/docs/System-IO.html#v:latin1

`encodeWith` would error on Chars outside of the range. There is very few, maybe no Encoding that is total under `encodeWith`, since you can represent Surrogates as Chars.

Reply:
> https://hackage.haskell.org/package/base-4.20.0.1/docs/System-IO.html#v:latin1

Thanks; I missed that `System.IO.TextEncoding` and `GHC.IO.Encoding.Types.TextEncoding` are the same. The latter is linked in the Haddocks in `System.OsString.Encoding`, e.g. [here](https://hackage.haskell.org/package/os-string-2.0.3/docs/System-OsString-Encoding.html#v:ucs2le).

Regarding the `FromJSON/ToJSON` instances: The more I think about it the less I am convince that they should be in this library in the first place - Or at least not brought in scope by `OsPath`. There is clearly more than one sensible instance and throughout the ecosystem that is dealt with by leaving the instance definition to the user of a library; For example there are no `Semigroup/Monoid` instances for `Int` for that reason.
It is of course convenient to have _some_ pre-defined instances at hand that cover most of of the use cases, but ideally those would be provided by another (public sub-) library. Personally I don't make use of those instance more often than not, and pulling in `aeson` as a dependency is in my projects often unnecessary. So ultimately I'd like to get rid of the (mandatory) `aeson` dependency at some point.

As for this PR, I'm inclined to move the JSON-related instances for `OsPath` to an own module, say `OsPath.Aeson`, and add a flag `aeson-instances` to `path` that controls whether any JSON-related instances are build at all (defaults to `true`).

Reply:
> As for this PR, I'm inclined to move the JSON-related instances for `OsPath` to an own module, say `OsPath.Aeson`, and add a flag `aeson-instances` to `path` that controls whether any JSON-related instances are build at all (defaults to `true`).

cabal flags must not control exposed API though, because a package can not specify "I need this flag of one of my dependencies enabled". Cabal has no means to specify this, so the end user may specify an invalid configuration and get a compile error with missing modules.

Reply:
Ah, yes, you are right indeed. And frankly I should've known better before making such a proposal, given how much time I spent trying to build GHCup before I figured that I have to set a flag for `streamly`... I'm quiet happy that this changed.

Reply:
We've had a discussion about aeson before:
https://github.com/commercialhaskell/path/issues/136

If there is no good aeson instance (for example because it needs IO) then I'd rather not provide any.